url6544f14c72_charles_d_burton_java_websocket_example_tgz-pJ8-minewebsocket_MineWebSocketJ8/src/minewebsocket/org/java_websocket/drafts/Draft_76.java

```diff
@@ -0,0 +1,242 @@
+				return HandshakeState.MATCHED;
+			} else {
+				return HandshakeState.NOT_MATCHED;
+			}
+		} catch ( InvalidHandshakeException e ) {
+			throw new RuntimeException( "bad handshakerequest", e );
+		}
+	}
+
+	@Override
+	public HandshakeState acceptHandshakeAsServer( ClientHandshake handshakedata ) {
+		if( handshakedata.getFieldValue( "Upgrade" ).equals( "WebSocket" ) && handshakedata.getFieldValue( "Connection" ).contains( "Upgrade" ) && handshakedata.getFieldValue( "Sec-WebSocket-Key1" ).length() > 0 && !handshakedata.getFieldValue( "Sec-WebSocket-Key2" ).isEmpty() && handshakedata.hasFieldValue( "Origin" ) )
+			return HandshakeState.MATCHED;
+		return HandshakeState.NOT_MATCHED;
+	}
+
+	@Override
+	public ClientHandshakeBuilder postProcessHandshakeRequestAsClient( ClientHandshakeBuilder request ) {
+		request.put( "Upgrade", "WebSocket" );
+		request.put( "Connection", "Upgrade" );
+		request.put( "Sec-WebSocket-Key1", generateKey() );
+		request.put( "Sec-WebSocket-Key2", generateKey() );
+
+		if( !request.hasFieldValue( "Origin" ) ) {
+			request.put( "Origin", "random" + reuseableRandom.nextInt() );
+		}
+
+		byte[] key3 = new byte[ 8 ];
+		reuseableRandom.nextBytes( key3 );
+		request.setContent( key3 );
+		return request;
+
+	}
+
+	@Override
+	public HandshakeBuilder postProcessHandshakeResponseAsServer( ClientHandshake request, ServerHandshakeBuilder response ) throws InvalidHandshakeException {
+		response.setHttpStatusMessage( "WebSocket Protocol Handshake" );
+		response.put( "Upgrade", "WebSocket" );
+		response.put( "Connection", request.getFieldValue( "Connection" ) ); // to respond to a Connection keep alive
+		response.put( "Sec-WebSocket-Origin", request.getFieldValue( "Origin" ) );
+		String location = "ws://" + request.getFieldValue( "Host" ) + request.getResourceDescriptor();
+		response.put( "Sec-WebSocket-Location", location );
+		String key1 = request.getFieldValue( "Sec-WebSocket-Key1" );
+		String key2 = request.getFieldValue( "Sec-WebSocket-Key2" );
+		byte[] key3 = request.getContent();
+		if( key1 == null || key2 == null || key3 == null || key3.length != 8 ) {
+			throw new InvalidHandshakeException( "Bad keys" );
+		}
+		response.setContent( createChallenge( key1, key2, key3 ) );
+		return response;
+	}
+
+	@Override
+	public Handshakedata translateHandshake( ByteBuffer buf ) throws InvalidHandshakeException {
+
+		HandshakeBuilder bui = translateHandshakeHttp( buf, role );
+		// the first drafts are lacking a protocol number which makes them difficult to distinguish. Sec-WebSocket-Key1 is typical for draft76
+		if( ( bui.hasFieldValue( "Sec-WebSocket-Key1" ) || role == Role.CLIENT ) && !bui.hasFieldValue( "Sec-WebSocket-Version" ) ) {
+			byte[] key3 = new byte[ role == Role.SERVER ? 8 : 16 ];
+			try {
+				buf.get( key3 );
+			} catch ( BufferUnderflowException e ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)